### PR TITLE
Fix hot restart flake

### LINKED ISSUE #121708: Mac web_tool_tests is 5.05% flaky
The post-submit test builder `Mac web_tool_tests` had a flaky ratio 5.05% for the past (up to) 100 commits, which is above our 2.00% threshold.

One recent flaky example for a same commit: https://ci.chromium.org/ui/p/flutter/builders/prod/Mac%20web_tool_tests/10805
Commit: https://github.com/flutter/flutter/commit/61759375652e6bd58e71c87a06ea43b726ad725f

Flaky builds:
https://ci.chromium.org/ui/p/flutter/builders/prod/Mac%20web_tool_tests/10805
https://ci.chromium.org/ui/p/flutter/builders/prod/Mac%20web_tool_tests/10799
https://ci.chromium.org/ui/p/flutter/builders/prod/Mac%20web_tool_tests/10794
https://ci.chromium.org/ui/p/flutter/builders/prod/Mac%20web_tool_tests/10758
https://ci.chromium.org/ui/p/flutter/builders/prod/Mac%20web_tool_tests/10755

Recent test runs:
https://flutter-dashboard.appspot.com/#/build?taskFilter=Mac%20web_tool_tests

Please follow https://github.com/flutter/flutter/wiki/Reducing-Test-Flakiness#fixing-flaky-tests to fix the flakiness and enable the test back after validating the fix (internal dashboard to validate: go/flutter_test_flakiness).

### PR DESCRIPTION
**Pull request by annagrin**
Fixes latest flakes in web hot restart tests. 

**Details**

I was able to add delays in various parts of the code to repro the flake pretty consistently - it was caused by a race between serving DevTools and hot restart with a following interleaving sequence of events:

- the call to https://github.com/flutter/flutter/blob/961df985faa4db9eff0245b3faa066fd0278cee9/packages/flutter_tools/lib/src/isolated/resident_web_runner.dart#L593 got the flutter views with the old isolate id in [`FlutterVmService.findExtensionIsolate`](https://github.com/flutter/flutter/blob/961df985faa4db9eff0245b3faa066fd0278cee9/packages/flutter_tools/lib/src/vmservice.dart#L979)
-  hot restart killed the isolate
-  chrome proxy service returned a sentinel which missed the check in https://github.com/flutter/flutter/blob/961df985faa4db9eff0245b3faa066fd0278cee9/packages/flutter_tools/lib/src/vmservice.dart#L1022  end errored instead in https://github.com/flutter/flutter/blob/961df985faa4db9eff0245b3faa066fd0278cee9/packages/flutter_tools/lib/src/vmservice.dart#L1025

**The fix**

Update the check to account for sentinel exceptions returned from `VmServerConnection` in https://github.com/dart-lang/sdk/blob/67f3e1f4a0dfe57c07192607095892a9d107a9d1/pkg/vm_service/lib/src/vm_service.dart#L1795

Closes: https://github.com/flutter/flutter/issues/121708

### CODE REVIEW

File: packages/flutter_tools/lib/src/vmservice.dart [hunk 1]
```diff
@@ -1019,7 +1019,10 @@ class FlutterVmService {
         onError: (Object? error, StackTrace stackTrace) {
           if (error is vm_service.SentinelException ||
             error == null ||
-            (error is vm_service.RPCError && error.code == RPCErrorCodes.kServiceDisappeared)) {
+            error is vm_service.RPCError &&
```
Comment (annagrin) [inline packages/flutter_tools/lib/src/vmservice.dart:1022]: @bkonyi chrome proxy service throws a sentinel exception on `getIsolate` for non-existing isolate id, as required by the VmServiceInterface, but the VmServerConnection translates it into an RPC error with RPCError.kInternalError as code, and the sentinel is added to the message as a string, so I have to check for the string contents here. Is there a better way to handle this error? Should we change the VMServer connection to throw something different, or have a set of error codes in `package:vm_service` we all agree on and can check for here?

File: packages/flutter_tools/lib/src/vmservice.dart [hunk 2]
```diff
@@ -1019,7 +1019,10 @@ class FlutterVmService {
...
+              (error.code == RPCErrorCodes.kServiceDisappeared ||
+                error.code == RPCErrorCodes.kInternalError &&
+                error.message.contains('Sentinel kind: Collected'))) {
```
Comment (annagrin) [inline packages/flutter_tools/lib/src/vmservice.dart:1025]: @bkonyi chrome proxy service throws a sentinel exception on getIsolate for non-existing isolate id, as required by the VmServiceInterface, but the VmServerConnection translates it into an RPC error with RPCError.kInternalError as code, and the sentinel is added to the message as a string, so I have to check for the string contents here. Is there a better way to handle this error? Should we change the VMServer connection to throw something different, or have a set of error codes in package:vm_service we all agree on and can check for here?

**Review by christopherfujino — APPROVED** [review]
LGTM

File: packages/flutter_tools/lib/src/vmservice.dart (hunk 2 above)
Comment (bkonyi) [reply packages/flutter_tools/lib/src/vmservice.dart:1025]: `service.getIsolate(...)` should be throwing a `SentinelException` in this case, so there's likely something wrong with the encoding on the DWDS side of things. What's the JSON response you're seeing here? It should look something like:

```json
{
  "id": "123",
  "result": {
    "type": "Sentinel",
    "kind": "Collected",
    "valueAsString": "..."
  }
}
```

Comment (annagrin) [reply packages/flutter_tools/lib/src/vmservice.dart:1025]: Chrome proxy service throws a sentinel exception, but the vm server connection we use to on top of it converts it to RPC error with code of `RPCError.kInternalError` here: https://github.com/flutter/flutter/blob/961df985faa4db9eff0245b3faa066fd0278cee9/packages/flutter_tools/lib/src/vmservice.dart#L1025

Should this code be changed then?

Comment (annagrin) [reply packages/flutter_tools/lib/src/vmservice.dart:1025]: I will file an issue in the SDK repo for this, in the meanwhile the current PR should fix the flake.

Comment (annagrin) [reply packages/flutter_tools/lib/src/vmservice.dart:1025]: Filed an issue (i am worried about all the other cases where we miss error checks due to this problem): https://github.com/dart-lang/sdk/issues/51752

**Comment by christopherfujino** [comment]
Landing on red luci-flutter as this fixes one of the reasons the tree is currently red

**Pull request merged**